src/System.Xml.XPath/src/System/Xml/XPath/Internal/BooleanFunctions.cs

```diff
@@ -60,7 +60,10 @@ internal bool toBoolean(XPathNodeIterator nodeIterator)
         {
             object result = _arg.Evaluate(nodeIterator);
             if (result is XPathNodeIterator) return _arg.Advance() != null;
-            if (result is string) return toBoolean((string)result);
+
+            string str = result as string;
+            if (str != null) return toBoolean(str);
+
             if (result is double) return toBoolean((double)result);
             if (result is bool) return (bool)result;
```

Comment:
if think you should change that too but I would rather write something like:
return result as string ?? result as double ?? result as bool ....